shelley/chain-and-ledger/executable-spec/src/PoolRank.hs

```diff
@@ -0,0 +1,93 @@
+module PoolRank
+  ( desirability
+  , rankedPools
+  ) where
+
+import           BaseTypes (intervalValue)
+import           Coin (Coin (..))
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import           Data.Ratio ((%))
+import qualified Data.Set as Set
+import           EpochBoundary (BlocksMade (..), SnapShots (..), Stake (..), poolStake)
+import           Keys (KeyHash)
+import           LedgerState (AccountState (..), EpochState (..), NewEpochState (..))
+import           Numeric.Natural (Natural)
+import           PParams (PParams (..))
+import           Slot (EpochSize)
+import           TxData (PoolParams (..))
+
+
+-- | Desirability calculation for non-myopic utily,
+-- corresponding to f^~ in section 5.6.1 of
+-- "Design Specification for Delegation and Incentives in Cardano"
+desirability
+  :: PParams
+  -> Coin
+  -> Natural
+  -> Natural
+  -> PoolParams crypto
+  -> Stake crypto
+  -> Coin
+  -> Rational
+desirability pp r blocksN blocksTotal pool (Stake stake) (Coin total) =
+  if fTilde <= cost
+    then 0
+    else (fTilde - cost)*(1 -margin)
+  where
+    fTilde      = fTildeNumer / fTildeDenom
+    fTildeNumer = pBar * fromIntegral r * (z0 + min s z0 * a0)
+    fTildeDenom = 1 + a0
+
+    cost   = fromIntegral $ _poolCost pool
+    margin = intervalValue $ _poolMargin pool
+    tot = max 1 (fromIntegral total)
```

Comment:
The `max 1` is to prevent division by zero, but will not be relevant once there is active stake, right?

Reply:
exactly